File: ga/glm4_result_2024_6/result/1968D_cpp_glm/37_1968D.cpp
use std::io::{self, BufRead};

fn calculate(p: &Vec<i32>, a: &Vec<i32>, pos: i32, k: i32) -> i64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];
    let mut pos = pos as usize;
    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = mx.max(cur + (k as i64) * (a[pos] as i64));
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }
    mx
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let t = lines.next().unwrap().unwrap().parse::<i32>().unwrap();

    for _ in 0..t {
        let mut line = lines.next().unwrap().unwrap();
        let mut iter = line.split_whitespace();
        let n = iter.next().unwrap().parse::<i32>().unwrap();
        let k = iter.next().unwrap().parse::<i32>().unwrap();
        let ps1 = iter.next().unwrap().parse::<i32>().unwrap();
        let ps2 = iter.next().unwrap().parse::<i32>().unwrap();

        let mut p = Vec::with_capacity(n as usize);
        let mut a = Vec::with_capacity(n as usize);

        for _ in 0..n {
            p.push(iter.next().unwrap().parse::<i32>().unwrap() - 1);
        }

        for _ in 0..n {
            a.push(iter.next().unwrap().parse::<i32>().unwrap());
        }

        let ans1 = calculate(&p, &a, ps1 - 1, k);
        let ans2 = calculate(&p, &a, ps2 - 1, k);

        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}

File: ga/4omini_result_2024_7/result/1968D_cpp_mini/37_1968D.rs
use std::io;

fn calculate(p: &Vec<usize>, a: &Vec<i64>, mut pos: usize, mut k: i64) -> i64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];

    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = mx.max(cur + k * a[pos]);
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }

    mx
}

fn main() {
    let stdin = io::stdin();
    let mut iter = stdin.lines();
    let t: usize = iter.next().unwrap().unwrap().trim().parse().unwrap();

    for _ in 0..t {
        // Reading inputs
        let first_line = iter.next().unwrap().unwrap();
        let mut params = first_line.split_whitespace();
        let n: usize = params.next().unwrap().parse().unwrap();
        let k: i64 = params.next().unwrap().parse().unwrap();
        let ps1: usize = params.next().unwrap().parse::<usize>().unwrap() - 1; // converting to 0-indexed
        let ps2: usize = params.next().unwrap().parse::<usize>().unwrap() - 1; // converting to 0-indexed

        let mut p: Vec<usize> = Vec::with_capacity(n);
        let mut a: Vec<i64> = Vec::with_capacity(n);

        // Read the p array
        let p_line = iter.next().unwrap().unwrap();
        for val in p_line.split_whitespace() {
            p.push(val.parse::<usize>().unwrap() - 1); // converting to 0-indexed
        }

        // Read the a array
        let a_line = iter.next().unwrap().unwrap();
        for val in a_line.split_whitespace() {
            a.push(val.parse::<i64>().unwrap());
        }

        // Calculate scores
        let ans1 = calculate(&p, &a, ps1, k);
        let ans2 = calculate(&p, &a, ps2, k);

        // Determine the result
        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}